src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -1634,29 +2054,27 @@ double ShenandoahFreeSet::internal_fragmentation() {
  *   d) Heap is half full, full and empty regions interleave => EF =~ 1
  */
 double ShenandoahFreeSet::external_fragmentation() {
-  size_t last_idx = 0;
+  idx_t last_idx = 0;
   size_t max_contig = 0;
   size_t empty_contig = 0;
-
   size_t free = 0;
-
-  for (size_t index = _free_sets.leftmost(Mutator); index <= _free_sets.rightmost(Mutator); index++) {
-    if (_free_sets.in_free_set(index, Mutator)) {
-      ShenandoahHeapRegion* r = _heap->get_region(index);
-      if (r->is_empty()) {
-        free += ShenandoahHeapRegion::region_size_bytes();
-        if (last_idx + 1 == index) {
-          empty_contig++;
-        } else {
-          empty_contig = 1;
-        }
+  idx_t rightmost = _partitions.rightmost(ShenandoahFreeSetPartitionId::Mutator);
+  for (idx_t index = _partitions.leftmost(ShenandoahFreeSetPartitionId::Mutator); index <= rightmost; ) {
+    assert(_partitions.in_free_set(ShenandoahFreeSetPartitionId::Mutator, index),
+           "Boundaries or find_first_set_bit failed: " SSIZE_FORMAT, index);
+    ShenandoahHeapRegion* r = _heap->get_region(index);
+    if (r->is_empty()) {
```

Comment:
This algorithm has changed from upstream. Is that intentional?

Reply:
That change came in when I merged from master.  I had assumed the change had also been applied upstream.  Is your diff compared to a "recent" upstream?
```
dc184f1099e0 (Kelvin Nilsen    2024-05-16 16:47:09 +0000 2037)     ShenandoahHeapRegion* r = _heap->get_region(index);
dc184f1099e0 (Kelvin Nilsen    2024-05-16 16:47:09 +0000 2038)     if (r->is_empty()) {
dc184f1099e0 (Kelvin Nilsen    2024-05-16 16:47:09 +0000 2039)       free += ShenandoahHeapRegion::region_size_bytes();
dc184f1099e0 (Kelvin Nilsen    2024-05-16 16:47:09 +0000 2040)       if (last_idx + 1 == index) {
dc184f1099e0 (Kelvin Nilsen    2024-05-16 16:47:09 +0000 2041)         empty_contig++;
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2042)       } else {
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2043)         empty_contig = 0;
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2044)       }
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2045)
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2046)       max_contig = MAX2(max_contig, empty_contig);
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2047)       last_idx = index;
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2048)     }
5f22805a50d4 (Kelvin Nilsen    2024-06-13 14:04:30 +0000 2049)     index = _partitions.find_index_of_next_available_region(Shenando\
ahFreeSetPartitionId::Mutator, index + 1);
f37149b1c510 (Aleksey Shipilev 2020-03-23 19:14:01 +0100 2050)   }
```

Reply:
I think we may have drifted from openjdk/jdk after we forked our generational shenandoah work?

Reply:
I must have messed something up during manual merge conflict resolution.
I'm fixing this code to match upstream.